frame/contracts/src/exec.rs

```diff
@@ -2097,4 +2116,87 @@ mod tests {
 			).unwrap();
 		});
 	}
+
+	#[test]
+	fn in_memory_changes_not_discarded() {
+		// Call stack: BOB -> CHARLIE (trap) -> BOB' (success)
+		// This tests verfies some edge case of the contract info cache:
+		// We change some value in our contract info before calling into a contract
+		// that calls into ourself. This triggers a case where BOBs contract info
+		// is written to storage and invalidated by the successful execution of BOB'.
+		// The trap of CHARLIE reverts the storage changes to BOB. When the root BOB regains
+		// control it reloads its contract info from storage. We check that changes that
+		// are made before calling into CHARLIE are not discarded.
+		let code_bob = MockLoader::insert(Call, |ctx, _| {
+			if ctx.input_data[0] == 0 {
+				let original_allowance = ctx.ext.rent_allowance();
+				let changed_allowance = <BalanceOf<Test>>::max_value() / 2;
+				assert_ne!(original_allowance, changed_allowance);
+				ctx.ext.set_rent_allowance(changed_allowance);
+				assert_eq!(
+					ctx.ext.call(0, CHARLIE, 0, vec![]).map(|v| v.0).map_err(|e| e.0),
```

Comment:
Nit: I would prefer if we reserved arguments of `assert` for non-side-effecting expressions.

Reply:
In production code I would agree. However, in tests there are a lot of asserts and enforcing this rule would result in many annoying variable bindings.